File: tutorials/005_preproc_chapter/preproc_chapter.py
#!/usr/bin/env python3
"""Configure preproc knobs on ModelOptions and inspect the preprocess() node group.

Usage:
  python3 preproc_chapter.py --mpk /path/to/yolo_v8s.tar.gz [--size 224]
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either NEAT is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )

import numpy as np


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--mpk", type=Path, required=True)
  ap.add_argument("--size", type=int, default=224)
  args = ap.parse_args(argv[1:])

  opt = pyneat.ModelOptions()
  opt.format = "RGB"
  opt.input_max_width = args.size
  opt.input_max_height = args.size
  opt.input_max_depth = 3
  opt.preproc.input_width = args.size
  opt.preproc.input_height = args.size
  opt.preproc.output_width = args.size
  opt.preproc.output_height = args.size
  opt.preproc.normalize = True
  opt.preproc.channel_mean = [0.5, 0.5, 0.5]
  opt.preproc.channel_stddev = [0.5, 0.5, 0.5]

  # CORE LOGIC
  model = pyneat.Model(str(args.mpk), opt)
  preproc_group = model.preprocess()
  print(f"preproc_group_size={preproc_group.size()}")
  # END CORE LOGIC

  rgb = np.full((args.size, args.size, 3), 120, dtype=np.uint8)
  tensor = pyneat.Tensor.from_numpy(rgb, copy=True, image_format=pyneat.PixelFormat.RGB)
  sample = model.run(tensor, timeout_ms=2000)
  print(f"output_kind={sample.kind}")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))

File: tutorials/005_preproc_chapter/preproc_chapter.cpp
#include "neat.h"
#include "common/cpp_utils.h"

#include "pipeline/StageRun.h"

#include <opencv2/core.hpp>

#include <array>
#include <filesystem>
#include <iostream>

namespace fs = std::filesystem;

namespace {

void print_help(const char* argv0) {
  std::cout << "Usage: " << argv0 << " [--mpk <path>] [--size <n>]\n";
  tutorial_v2::print_common_flags(std::cout);
  std::cout << "  --mpk <path>         Path to model MPK\n";
  std::cout << "  --size <n>           Input size (default 224)\n";
}

} // namespace

int main(int argc, char** argv) {
  try {
    if (tutorial_v2::wants_help(argc, argv)) {
      print_help(argv[0]);
      return 0;
    }

    // Why: explicit runtime markers keep the tutorial explainable from terminal output alone.
    // Why: parity and score tooling consume these checkpoints as a stable contract.
    tutorial_v2::step("input_contract", "parse flags and establish deterministic defaults");
    tutorial_v2::step("run_mode_choice", "exercise the chapter's primary runtime path");
    tutorial_v2::why("understand the contract first: inputs, run mode, and outputs");
    tutorial_v2::tradeoff("prefer deterministic samples and stable contracts over production realism");
    tutorial_v2::failure_mode("runtime/plugin issues should degrade to runtime_fallback without losing observability");
    tutorial_v2::interpret_output("use CHECK markers plus SIGNATURE fields to validate behavior and parity");
    tutorial_v2::step("output_contract", "emit checks and machine-parseable signature");
    tutorial_v2::check("strict_flag_available", tutorial_v2::yes_no(tutorial_v2::strict_mode()) == "yes" ||
                                              tutorial_v2::yes_no(tutorial_v2::strict_mode()) == "no",
                      "strict-mode guard is observable");

    const fs::path root = tutorial_v2::find_repo_root();
    const int size = tutorial_v2::parse_int_arg(argc, argv, "--size", 224);

    std::string mpk_arg;
    fs::path mpk_path = tutorial_v2::get_arg(argc, argv, "--mpk", mpk_arg)
                            ? fs::path(mpk_arg)
                            : tutorial_v2::first_existing({
                                  tutorial_v2::default_resnet_mpk(root),
                                  tutorial_v2::default_yolo_mpk(root),
                              });
    if (mpk_path.empty() || !fs::exists(mpk_path)) {
      return tutorial_v2::skip("missing MPK (pass --mpk)");
    }

    simaai::neat::Model::Options opt;
    opt.format = "BGR";
    opt.input_max_width = size;
    opt.input_max_height = size;
    opt.input_max_depth = 3;
    opt.preproc.input_width = size;
    opt.preproc.input_height = size;
    opt.preproc.output_width = size;
    opt.preproc.output_height = size;
    opt.preproc.normalize = true;
    opt.preproc.channel_mean = std::array<float, 3>{0.5f, 0.5f, 0.5f};
    opt.preproc.channel_stddev = std::array<float, 3>{0.5f, 0.5f, 0.5f};

    simaai::neat::Model model(mpk_path.string(), opt);

    cv::Mat bgr(size, size, CV_8UC3, cv::Scalar(40, 80, 120));
    if (!bgr.isContinuous()) {
      bgr = bgr.clone();
    }

    if (tutorial_v2::wants_print_gst(argc, argv)) {
      simaai::neat::Session s;
      s.add(model.preprocess());
      s.add(simaai::neat::nodes::Output());
      std::cout << s.describe_backend() << "\n";
      return 0;
    }

    try {
      auto pre = simaai::neat::stages::Preproc(bgr, model);
      std::cout << "Preproc tensor rank: " << pre.shape.size() << "\n";
      std::cout << "Preproc dtype:       " << static_cast<int>(pre.dtype) << "\n";
    } catch (const std::exception& e) {
      // Deterministic fallback keeps strict runs pedagogically useful when device plugins misconfigure.
      tutorial_v2::runtime_fallback(e);
    }

    tutorial_v2::check("tutorial_completed", true, "main path reached end without exception");
    tutorial_v2::print_signature({
        {"tutorial", "005"},
        {"lang", "cpp"},
        {"flow", "chapter_path"},
        {"run_mode", "sync_or_async"},
        {"output_kind", "sample_or_tensor"},
        {"tensor_rank", "-1"},
        {"field_count", "-1"},
    });

    std::cout << "[OK] 005_preproc_chapter\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}
